### PR TITLE
First cut at the core sync infrastructure

### PR DESCRIPTION
**Pull request by mhammond**
Builds on #2810. This needs the trait from Lina's patch, but is setup to work with it. It implements the 3-way merge (although doesn't quite get deletions right yet, so there's still some work to do) - but comments welcome.

### CODE REVIEW

**Comment by thomcc** [comment]
Regarding how to do the two/three way merge, I'm pretty happy with how https://github.com/mozilla/application-services/blob/master/components/logins/src/update_plan.rs#L34-L60 works, and it's a high level description of the set of operations you perform for a three way merge.

https://github.com/mozilla/application-services/blob/master/components/logins/src/db.rs#L1061-L1114 is the code that calls it, and decides how to handle deletions vs missing mirror, etc. I think trying to handle it as you have is going to be tricky (but your comment says just about as much).

That said, handing things in bulk as this does is clearly not necessary, it could still be done piecemeal.

File: components/webext-storage/src/sync/incoming.rs [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+/// records. Which one depends on the IncomingState.
+#[derive(Debug, PartialEq)]
+pub enum IncomingAction {
+    // Something's wrong with this entry - probably JSON?
+    // (but we seem to be getting away without this for now)
+    //Invalid { reason: String },
+    /// We should locally delete the data for this record
+    DeleteLocally,
+    /// We should remotely delete the data for this record
+    DeleteRemotely,
```
Comment (mhammond) [inline components/webext-storage/src/sync/incoming.rs:153]: incoming should never cause remote deletion

File: components/webext-storage/src/sync/incoming.rs [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+                    // Eg, final key removed remotely, different key added
+                    // locally, the new key should still be added.
+                    IncomingAction::DeleteLocally
+                }
+            }
+        }
+        IncomingState::LocalOnly { incoming, local } => {
+            // So we have a local record and an incoming/staging record, but *not* a
+            // mirror record. This is the first time we've seen this (ie, almost
+            // certainly another device synced something)
```
Comment (mhammond) [inline components/webext-storage/src/sync/incoming.rs:196]: another *first* synced something

File: components/webext-storage/src/sync/incoming.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+            // we've ever seen it. No conflict possible, just take the remote.
+            match incoming {
+                Some(data) => IncomingAction::TakeRemote { data },
+                None => IncomingAction::DeleteLocally,
+            }
+        }
+    }
+}
+
+pub fn apply_actions<S: ?Sized + Interruptee>(
```
Comment (mhammond) [inline components/webext-storage/src/sync/incoming.rs:239]: comments

File: components/webext-storage/src/sync/outgoing.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        signal.err_if_interrupted()?;
+        conn.execute_named(
+            sql,
+            rusqlite::named_params! {
+                ":old_counter": state.change_counter,
+                ":sync_status_normal": SyncStatus::Normal as u8,
+                ":ext_id": state.ext_id,
+            },
+        )?;
+    }
```
Comment (rfk) [inline components/webext-storage/src/sync/outgoing.rs:98]: I understand this code isn't all plumbed together yet, but would I be correct to expect that this function should also update the mirror to reflect what was uploaded to the server? Or, will we sync down our own changes and only update the mirror at that point?

File: components/webext-storage/sql/create_schema.sql [hunk 5]
```diff
@@ -5,19 +5,20 @@
...
+    sync_change_counter INTEGER NOT NULL DEFAULT 1,
 
-CREATE TABLE IF NOT EXISTS moz_extension_data_tombstones (
-    guid TEXT PRIMARY KEY
+    -- We only keep tombstones for items with a sync_status of Normal (and only
+    -- then until we upload the tombstone to the server. Enforce that.
+    CHECK(sync_status = 2 OR data IS NOT NULL)
 ) WITHOUT ROWID;
 
+/* index on sync_status? */
```
[the remark this answers is not in the document]
Comment (rfk) [reply components/webext-storage/sql/create_schema.sql:20]: That has the potential to be a very churny index, basically every write to this table will also need to update the index. It could be better on balance to pay the full table scan on each sync (not entirely sure how to test the two options in practice tho)

File: components/webext-storage/src/sync/outgoing.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        let raw_data: Option<String> = row.get("data")?;
+        let (data, deleted) = if raw_data.is_some() {
+            (raw_data, false)
+        } else {
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
+                ext_id: ext_id.clone(),
+                change_counter: row.get("sync_change_counter")?,
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/outgoing.rs:45]: yes, exactly that. It could be a bool if the entire sync happened in a single transaction. Instead, we track what the value was when starting, then decrement it by that later. This doesn't have all the change counter stuff wired up (a later version is far more complete)

File: components/webext-storage/src/sync/outgoing.rs (hunk 4 above)
Comment (mhammond) [reply components/webext-storage/src/sync/outgoing.rs:98]: yes, exactly. That's in a later version too.

File: components/webext-storage/src/sync/incoming.rs [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+            match (incoming, local) {
+                (Some(id), Some(ld)) => {
+                    // This means the extension exists locally and remotely
+                    // but this is the first time we've synced it. That's no problem, it's
+                    // just a 2-way merge...
+                    merge(id, ld, None)
+                }
+                (Some(_), None) => {
+                    // We've data locally, but there's an incoming deletion.
+                    // Remote wins.
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/incoming.rs:206]: Yes - what's on the server always wins. However, it's not clear what the exact semantics are here - see below.

File: components/webext-storage/src/sync/mod.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    #[serde(default, skip_serializing_if = "is_default")]
+    deleted: bool,
+    last_modified: ServerTimestamp,
+}
+
+// Perform a 2-way or 3-way merge, where the incoming value wins on confict.
+// XXX - this needs more thought, and probably needs significant changes.
+// Main problem is that it doesn't handle deletions - but to do that, we need
+// something other than a simple Option<JsonMap> - we need to differentiate
+// "doesn't exist" from "removed".
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/mod.rs:45]: Yes, we could do this, but TBH I'm not sure what the exact semantics are here - it's not documented, so we probably need to check how chrome behaves in these cases. I've just punted doing anything else here for now as hopefully @glasserc is tackling that as part of #2900

File: components/webext-storage/src/sync/incoming.rs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+            }
+
+            // We merged this data, so need to update locally but still consider
+            // it dirty because the merged data must be uploaded.
+            IncomingAction::Merge { data } => {
+                tx.execute_named_cached(
+                    "UPDATE storage_sync_data SET data = :data, sync_change_counter = sync_change_counter + 1 WHERE ext_id = :ext_id",
+                    &[
+                        (":ext_id", &item.ext_id),
+                        (":data", &serde_json::Value::Object(data).to_string()),
```
Comment (linacambridge) [inline components/webext-storage/src/sync/incoming.rs:271]: Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature? Paging @thomcc to make sure! ☎️

File: components/webext-storage/src/sync/incoming.rs [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+    use super::*;
+    use crate::api;
+    use crate::db::test::new_mem_db;
+    use interrupt::NeverInterrupts;
+    use rusqlite::NO_PARAMS;
+    use serde_json::{json, Value};
+
+    // select simple int
+    fn ssi(conn: &Connection, stmt: &str) -> u32 {
+        conn.query_row_and_then(stmt, rusqlite::NO_PARAMS, |row| row.get::<_, u32>(0))
```
Comment (linacambridge) [inline components/webext-storage/src/sync/incoming.rs:300]: I know this is test code, but `try_query_one::<u32>` might replace this, too.

File: components/webext-storage/src/sync/incoming.rs [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+            IncomingState::Everywhere {
+                incoming: None,
+                local: None,
+                mirror: None,
+            }
+        );
+        Ok(())
+    }
+
+    // XXX - test apply_actions!
```
Comment (linacambridge) [inline components/webext-storage/src/sync/incoming.rs:469]: Yes, please! 😁 Thanks also for the wonderful tests so far!

File: components/webext-storage/src/sync/mod.rs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+/// For use with `#[serde(skip_serializing_if = )]`
+#[inline]
+pub fn is_default<T: PartialEq + Default>(v: &T) -> bool {
+    *v == T::default()
+}
+
+// XXX - need to work out how to consolidate "payload" vs "bso", particularly
+// the timestamp, etc. For now, we represent this in ServerPayload, even though
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
```
Comment (linacambridge) [inline components/webext-storage/src/sync/mod.rs:29]: Hmmm, yeah, we're going to need to pass that over in Desktop. I wonder if we can "just" pass the raw decrypted JSON over XPCOM, and have the Desktop integration layer parse it into a `sync15_traits::Payload` (which has `id`, `deleted`, and `data` already on it) before handing it over to `webext_storage`.

File: components/webext-storage/src/sync/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (linacambridge) [reply components/webext-storage/src/sync/mod.rs:45]: Summarizing the discussion from that thread, it sounds like Chrome's behavior is too coarse, and not really what we want...but also, the Kinto "client wins" strategy isn't right for us, and it's more important to maintain compatibility with what we had before than with Chrome's implementation. Is that accurate?

Since we have the mirror copy, @rfk's suggestion makes sense!

File: components/webext-storage/src/sync/mod.rs [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    #[test]
+    fn test_3way_merging() -> Result<()> {
+        // No conflict - identical local and remote.
+        assert_eq!(
+            merge(
+                map!({"one": "one", "two": "two"}),
+                map!({"two": "two", "one": "one"}),
+                Some(map!({"parent_only": "parent"})),
```
Comment (linacambridge) [inline components/webext-storage/src/sync/mod.rs:116]: Hmm, in this case, we still want to update the mirror without reuploading, though, no? (Or maybe it's fine to leave it as is, if it can't cause problems for future merges...say, if local or remote ever brings `parent_only` back).

Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

File: components/webext-storage/src/sync/outgoing.rs [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            .unwrap_or_else(SyncGuid::random);
+        let ext_id: String = row.get("ext_id")?;
+        let raw_data: Option<String> = row.get("data")?;
+        let (data, deleted) = if raw_data.is_some() {
+            (raw_data, false)
+        } else {
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
```
Comment (linacambridge) [inline components/webext-storage/src/sync/outgoing.rs:47]: Hmmm, this does mean we'll need the Sync JS side to hold on to this, match them up before calling `setUploaded`, and pass them back. That should work also, but how do you feel about storing all these outgoing ones in a temp table, like Places does it? `guid` and `change_counter` would become columns in that table, and when you mark it as uploaded, doing the write back to the mirror?

File: components/webext-storage/src/sync/incoming.rs [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+pub struct IncomingItem {
+    guid: SyncGuid,
+    ext_id: String,
+}
+
+/// The "state" we find ourselves in when considering an incoming/staging
+/// record. This "state" is the input to calculating the IncomingAction.
+#[derive(Debug, PartialEq)]
+pub enum IncomingState {
+    IncomingOnly {
```
[the remark this answers is not in the document]
Comment (glasserc) [reply components/webext-storage/src/sync/incoming.rs:77]: I'd also like to express that the types themselves are confusing to me. I would have assumed that e.g. `IncomingOnly` would require a non-`Option` `JsonMap`. The fact that all of these are `Options` in all variants makes me wonder what the difference is between a simple `struct IncomingState { incoming: Option<JsonMap>, local: Option<JsonMap>, mirror: Option<JsonMap> }`. I'm guessing that the use of `Option` here doesn't necessarily mean "missing" but means something more like tombstones?

File: components/webext-storage/src/sync/mod.rs [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ServerPayload {
+    guid: SyncGuid,
+    ext_id: String,
+    #[serde(default, skip_serializing_if = "is_default")]
+    data: Option<String>,
+    #[serde(default, skip_serializing_if = "is_default")]
+    deleted: bool,
```
Comment (glasserc) [inline components/webext-storage/src/sync/mod.rs:37]: What does it mean if `deleted = false` but `data = Nothing`? Is this the same as an empty object?

File: components/webext-storage/src/sync/incoming.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/incoming.rs:77]: yes, exactly, it means a tombstone. I can probably make that more explicit by using yet another enum to make it more explicit. I agree the distinction between an empty object and an explicit tombstone isn't clear either.

TBH though, I've been kinda deferring this until this discussion about the semantics.  IIUC, if we went with Chrome's observed behaviour, we wouldn't even need tombstones, but because we are cloning the existing impl, we do, right?

How does the current impl work in this regard? Is there any distinction between an explicit deletion vs an empty object? But as above, I'm not tied to anything here in any way, as I was just ignoring it :)

[the remark this answers is not in the document]
Comment (glasserc) [reply components/webext-storage/src/sync/incoming.rs:77]: My understanding of Chrome's behavior is that the client has tombstones but the server doesn't. So I think we would still need to represent tombstones to *some* extent. The current implementation does have tombstones both on the client and on the server. Actually, the current implementation has tombstones on a per-key basis, which I don't think we can easily represent given the structures we have now, but maybe I'm missing something.

File: components/webext-storage/src/sync/mod.rs (hunk 16 above)
Comment (mhammond) [reply components/webext-storage/src/sync/mod.rs:37]: Yeah, it is representing the actual format on the server. I'm still not 100% happy that's the right choice for the server either though - how is this redundant and potentially ambiguous  option better than an explicit data=null?

File: components/webext-storage/src/sync/incoming.rs [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+            ],
+        )?;
+    }
+    tx.commit()?;
+    Ok(())
+}
+
+/// Details about an incoming item.
+#[derive(Debug, PartialEq)]
+pub struct IncomingItem {
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/incoming.rs:68]: I ended up playing with, but not doing this. I was a bit torn as the 3 structs and various tuples does seem a bit clunky at first glance, but:

* `plan_incoming` consumes an `IncomingState` and produces an `IncomingAction` - it doesn't need to know the guid or ext-id. If the `IncomingItem` was part of the state, I'd need to work out how to thread things so the item came out of `plan_incoming()` as part of consuming the state - which would be a bit unnatural as it's not needed there.

* `merge()` makes this even more complicated - it just takes json as inputs and creates an `IncomingAction` as output - which `plan_incoming()` just hands back to its caller - so threading the item all the way through `merge()` makes things quite ugly.

The tl;dr is that in the current setup, all functions use exactly what they need and nothing more. I'm happy to revisit this later though, particularly if Ethan's work makes some of the above no longer true/relevant.

File: components/webext-storage/src/sync/incoming.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mhammond) [reply components/webext-storage/src/sync/incoming.rs]: I ended up adding another Enum -

```
/// The state data can be in. Could be represented as Option<JsonMap>, but this
/// is clearer and independent of how the data is stored.
#[derive(Debug, PartialEq)]
pub enum DataState {
    /// The data was deleted.
    Deleted,
    /// Data exists, as stored in the map.
    Exists(JsonMap),
}
```

Which is more verbose, but does make the various states easier to read. I pushed it in rev e478c2142248ad89cf638d0b35d326a45b8101aa - so it's easy to see the changes, but also easy to back out if anyone things it's a backwards step - LMKWYT!

File: components/webext-storage/src/sync/mod.rs (hunk 13 above)
Comment (mhammond) [reply components/webext-storage/src/sync/mod.rs:134]: > Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

Yep - the final outgoing process writes staging to the mirror.

File: components/webext-storage/src/sync/incoming.rs (hunk 9 above)
Comment (mhammond) [reply components/webext-storage/src/sync/incoming.rs:310]: > Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature?

It does indeed - nice catch!

**Comment by mhammond** [comment]
phew! Thanks the the thorough review @linacambridge! I think I've addressed most review comments here - what's outstanding is:

* Working out what the payload struct should look like - I figured it's best to do that as part of wiring up the sync bridge, and it should be quite easy to retrofit changes here.

* A few outstanding items around merging, for which I'm going to wait for Ethan.

File: components/webext-storage/src/sync/incoming.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (linacambridge) [reply components/webext-storage/src/sync/incoming.rs:84]: That sounds perfect to me, thanks for taking a look and the thorough response! I agree that the alternatives are even worse, and the way you have this set up now is the cleanest.

File: components/webext-storage/src/sync/incoming.rs [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+    guid: SyncGuid,
+    ext_id: String,
+}
+
+/// The "state" we find ourselves in when considering an incoming/staging
+/// record. This "state" is the input to calculating the IncomingAction.
+#[derive(Debug, PartialEq)]
+pub enum IncomingState {
+    /// There's an incoming item, but data for that extension doesn't exist
```
Comment (linacambridge) [inline components/webext-storage/src/sync/incoming.rs:93]: I love these comments, thank you so much for adding them! 😍

File: components/webext-storage/src/sync/incoming.rs [hunk 19]
```diff
@@ -0,0 +1,648 @@
...
+    /// either in our local data store or in the local mirror. IOW, this is the
+    /// very first time we've seen this extension.
+    IncomingOnly { incoming: DataState },
+    /// There's an incoming item and we have data for the same extension in
+    /// our local store - but not in our mirror. This should be relatively
+    /// uncoming as it means:
```
Comment (linacambridge) [inline components/webext-storage/src/sync/incoming.rs:99]: Nit: uncommon

File: components/webext-storage/src/sync/mod.rs (hunk 12 above)
Comment (linacambridge) [reply components/webext-storage/src/sync/mod.rs:29]: I spent a bit of time noodling on this, and the result is #3067 and https://phabricator.services.mozilla.com/D73581. Basically, we have this thing called an "envelope", which is BSO metadata fields + the unparsed payload. An incoming envelope has a `payload()` method, which you can use to get at the payload—which you can then either deserialize into a record, or check if it's a tombstone via `is_deleted`, like you can in a-s `sync15`.

I think we can land your patch as-is now, and, assuming you like the envelope idea, go back and tweak this after.

File: components/webext-storage/src/sync/outgoing.rs [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+    tx: &Transaction<'_>,
+    _signal: &dyn Interruptee,
+) -> Result<Vec<ServerPayload>> {
+    // The item may not yet have a GUID (ie, it might not already be in either
+    // the mirror nor the incoming staging table.) If not for this fact, we
+    // could probably have arranged for the populating of the outgoing staging
+    // table to be done via a single `INSERT INTO ... SELECT` statement - but
+    // due to the GUID complication, and the fact we need to extract each
+    // record from this query anyway means we just do a single query, then
+    // perform the inserts one row at a time.
```
Comment (linacambridge) [inline components/webext-storage/src/sync/outgoing.rs:60]: I _think_ you can still do it like...

```sql
INSERT INTO storage_sync_outgoing_staging
(guid, ext_id, data, sync_change_counter)
SELECT coalesce(m.guid, s.guid, generate_guid()),
       l.ext_id, l.data, l.sync_change_counter
FROM storage_sync_data l
LEFT JOIN storage_sync_mirror m ON m.ext_id = l.ext_id
LEFT JOIN storage_sync_staging s ON s.ext_id = l.ext_id
WHERE sync_change_counter > 0
```

...And then read everything out of the temp table, but you're right—since we still need to return a `Vec<...>` of them, anyway, we may as well do it in Rust, and read it from the durable tables before staging! 👍

File: components/webext-storage/src/sync/outgoing.rs [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    tx: &Transaction<'_>,
+    items: &[SyncGuid],
+    signal: &dyn Interruptee,
+) -> Result<()> {
+    log::debug!(
+        "record_uploaded recording that {} items were uploaded",
+        items.len()
+    );
+
+    let sql = "
```
Comment (linacambridge) [inline components/webext-storage/src/sync/outgoing.rs:106]: I was gonna suggest doing this in a trigger, but we can do it in a follow-up, or leave it be.

File: components/webext-storage/src/sync/outgoing.rs [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+    tx.execute_batch(
+        "
+        INSERT OR REPLACE INTO storage_sync_mirror (guid, ext_id, data)
+        SELECT guid, ext_id, data FROM temp.storage_sync_staging;
+
+        INSERT OR REPLACE INTO storage_sync_mirror (guid, ext_id, data)
+        SELECT guid, ext_id, data FROM temp.storage_sync_outgoing_staging;
+
+        -- XXX - is this correct?
+        DELETE from storage_sync_data WHERE data IS NULL AND sync_change_counter = 0;",
```
Comment (linacambridge) [inline components/webext-storage/src/sync/outgoing.rs:139]: I think so—durable tombstones get stored in the mirror, so there's no need for them to hang around locally once we've synced them.
